src/acl_platform.cpp

```diff
@@ -737,6 +742,25 @@ static void l_add_device(int idx) {
   device->address_bits = 64;              // Yes, our devices are 64-bit.
 }
 
+void *l_eagerly_update_device_op_queue(void *arg) {
+  while (true) {
+    std::scoped_lock lock{acl_mutex_wrapper};
+
+    // Sleep if no interrupt happening
+    acl_wait_for_device_update(NULL);
```

Comment:
The side effect of using this existing call is that the thread gets woken up not just when there is a mmd triggered kernel interrupt, but also when all the signal handler functions are called (`acl_receive_kernel_update`, `acl_set_execution_status`, `acl_receive_device_exception`, and `acl_schedule_printf_buffer_pickup`). But this is already better then always polling, I wonder if it is necessary to fine grain this even further (might be risky)?

Reply:
Can potentially
- Introduce a new `acl_condvar_s`, or
- Use a semaphore (is this doable?)
To signal to the device op queue update thread, to avoid waking up from event status update calls, etc. 

However, the device op queue update thread might still need to acquire the global lock when doing the update, so that no other thread is modifying the shared resources (which are quite a lot...).

Reply:
Thanks for the feedback! Yes, the primary goal is to make forward progress in processing the device operation queue. In the model you suggested, is it kind of like treating the device op queue as a buffer for consumer (runtime spawned thread) and producer (user threads)? This conceptually makes sense, just that I'm not sure if there are other runtime constructs (events, command queue, etc.) modified during the process of updating the device op queue, would that pose any obstacle to the consumer-producer model?

Also I'm thinking, if at the end of the day we are actively making progress on the device op queue, maybe it would be possible to remove some of the `acl_idle_updates` calls in `clEnqueue...` used to nudge the device op queue. I'm not sure if this idea is practical in any sense or if it will cause any issue though...